File: packages/agent-cache/src/AgentCache.ts
import type {
  Valkey,
  AgentCacheOptions,
  AgentCacheStats,
  ToolEffectivenessEntry,
  ToolRecommendation,
  TierStats,
  SessionStats,
  ToolStats,
  ModelCost,
} from './types';
import { DEFAULT_COST_TABLE } from './defaultCostTable';
import { LlmCache } from './tiers/LlmCache';
import { ToolCache } from './tiers/ToolCache';
import { SessionStore } from './tiers/SessionStore';
import { createTelemetry } from './telemetry';
import { createAnalytics, NOOP_ANALYTICS, type Analytics } from './analytics';
import { ValkeyCommandError } from './errors';
import { escapeGlobPattern } from './utils';
import { clusterScan } from './cluster';

export class AgentCache {
  public readonly llm: LlmCache;
  public readonly tool: ToolCache;
  public readonly session: SessionStore;

  private readonly client: Valkey;
  private readonly name: string;
  private readonly statsKey: string;
  private readonly defaultTtl: number | undefined;
  private readonly toolTierTtl: number | undefined;
  private analytics: Analytics = NOOP_ANALYTICS;
  private statsTimer: ReturnType<typeof setInterval> | undefined;
  private shutdownCalled = false;

  constructor(options: AgentCacheOptions) {
    this.client = options.client;
    this.name = options.name ?? 'betterdb_ac';
    this.statsKey = `${this.name}:__stats`;
    this.defaultTtl = options.defaultTtl;
    this.toolTierTtl = options.tierDefaults?.tool?.ttl;

    const telemetry = createTelemetry({
      prefix: options.telemetry?.metricsPrefix ?? 'agent_cache',
      tracerName: options.telemetry?.tracerName ?? '@betterdb/agent-cache',
      registry: options.telemetry?.registry,
    });

    const defaultTtl = options.defaultTtl;

    const useDefault = options.useDefaultCostTable ?? true;
    const effectiveCostTable: Record<string, ModelCost> | undefined = useDefault
      ? { ...DEFAULT_COST_TABLE, ...(options.costTable ?? {}) }
      : options.costTable;

    this.llm = new LlmCache({
      client: this.client,
      name: this.name,
      defaultTtl,
      tierTtl: options.tierDefaults?.llm?.ttl,
      costTable: effectiveCostTable,
      telemetry,
      statsKey: this.statsKey,
    });

    this.tool = new ToolCache({
      client: this.client,
      name: this.name,
      defaultTtl,
      tierTtl: options.tierDefaults?.tool?.ttl,
      telemetry,
      statsKey: this.statsKey,
    });

    this.session = new SessionStore({
      client: this.client,
      name: this.name,
      defaultTtl,
      tierTtl: options.tierDefaults?.session?.ttl,
      telemetry,
      statsKey: this.statsKey,
    });

    // Fire-and-forget: load persisted tool policies from Valkey
    this.tool.loadPolicies().catch(() => {});

    // Fire-and-forget: initialize product analytics
    const analyticsOpts = options.analytics;
    createAnalytics({
      apiKey: analyticsOpts?.apiKey,
      host: analyticsOpts?.host,
      disabled: analyticsOpts?.disabled,
    })
      .then((a) => {
        if (this.shutdownCalled) {
          // If shutdown won the race, ensure the late-created analytics client
          // is also torn down so its internal timers do not keep the process alive.
          return a.shutdown();
        }
        this.analytics = a;
        const configProps: Record<string, unknown> = {
          defaultTtl: options.defaultTtl,
          llmTtl: options.tierDefaults?.llm?.ttl,
          toolTtl: options.tierDefaults?.tool?.ttl,
          sessionTtl: options.tierDefaults?.session?.ttl,
          hasCostTable: !!options.costTable,
          usesDefaultCostTable: useDefault,
        };
        return a.init(this.client, this.name, configProps);
      })
      .then(() => {
        if (this.shutdownCalled) return;
        const intervalMs = analyticsOpts?.statsIntervalMs ?? 300_000;
        if (intervalMs > 0) {
          this.statsTimer = setInterval(() => this.captureStatsSnapshot(), intervalMs);
          this.statsTimer.unref();
        }
      })
      .catch(() => {});
  }

  async stats(): Promise<AgentCacheStats> {
    let raw: Record<string, string>;
    try {
      raw = await this.client.hgetall(this.statsKey) ?? {};
    } catch (err) {
      throw new ValkeyCommandError('HGETALL', err);
    }

    const getInt = (field: string): number => {
      const val = raw[field];
      return val ? parseInt(val, 10) : 0;
    };

    const computeHitRate = (hits: number, misses: number): number => {
      const total = hits + misses;
      return total > 0 ? hits / total : 0;
    };

    // LLM tier stats
    const llmHits = getInt('llm:hits');
    const llmMisses = getInt('llm:misses');
    const llmTotal = llmHits + llmMisses;
    const llmStats: TierStats = {
      hits: llmHits,
      misses: llmMisses,
      total: llmTotal,
      hitRate: computeHitRate(llmHits, llmMisses),
    };

    // Tool tier stats
    const toolHits = getInt('tool:hits');
    const toolMisses = getInt('tool:misses');
    const toolTotal = toolHits + toolMisses;
    const toolStats: TierStats = {
      hits: toolHits,
      misses: toolMisses,
      total: toolTotal,
      hitRate: computeHitRate(toolHits, toolMisses),
    };

    // Session stats
    const sessionStats: SessionStats = {
      reads: getInt('session:reads'),
      writes: getInt('session:writes'),
    };

    // Cost saved
    const costSavedMicros = getInt('cost_saved_micros');

    // Per-tool stats
    const perTool: Record<string, ToolStats> = {};
    const toolPattern = /^tool:([^:]+):(hits|misses|cost_saved_micros)$/;

    for (const [key, value] of Object.entries(raw)) {
      const match = key.match(toolPattern);
      if (match) {
        const toolName = match[1];
        const statType = match[2];
        const numValue = parseInt(value, 10);

        if (!perTool[toolName]) {
          perTool[toolName] = {
            hits: 0,
            misses: 0,
            hitRate: 0,
            ttl: this.tool.getPolicy(toolName)?.ttl,
            costSavedMicros: 0,
          };
        }

        if (statType === 'hits') {
          perTool[toolName].hits = numValue;
        } else if (statType === 'misses') {
          perTool[toolName].misses = numValue;
        } else if (statType === 'cost_saved_micros') {
          perTool[toolName].costSavedMicros = numValue;
        }
      }
    }

    // Compute hit rates for per-tool stats
    for (const perToolEntry of Object.values(perTool)) {
      perToolEntry.hitRate = computeHitRate(perToolEntry.hits, perToolEntry.misses);
    }

    return {
      llm: llmStats,
      tool: toolStats,
      session: sessionStats,
      costSavedMicros,
      perTool,
    };
  }

  async toolEffectiveness(): Promise<ToolEffectivenessEntry[]> {
    // Reuse data already fetched by stats() to avoid N+1 queries
    const stats = await this.stats();
    const entries: ToolEffectivenessEntry[] = [];

    for (const [toolName, toolStats] of Object.entries(stats.perTool)) {
      // Cost saved is already computed in perTool from the single HGETALL call (microdollars -> dollars)
      const costSaved = toolStats.costSavedMicros / 1_000_000;

      // Resolve effective TTL through full hierarchy: policy -> tierTtl -> defaultTtl
      const policyTtl = this.tool.getPolicy(toolName)?.ttl;
      const effectiveTtl = policyTtl ?? this.toolTierTtl ?? this.defaultTtl;

      // Generate recommendation based on hit rate and effective TTL
      let recommendation: ToolRecommendation;

      if (toolStats.hitRate > 0.8) {
        // High hit rate - consider increasing TTL (unless already > 1 hour or no TTL)
        if (effectiveTtl !== undefined && effectiveTtl < 3600) {
          recommendation = 'increase_ttl';
        } else {
          recommendation = 'optimal';
        }
      } else if (toolStats.hitRate >= 0.4) {
        recommendation = 'optimal';
      } else {
        recommendation = 'decrease_ttl_or_disable';
      }

      entries.push({
        tool: toolName,
        hitRate: toolStats.hitRate,
        costSaved,
        recommendation,
      });
    }

    // Sort by costSaved descending
    entries.sort((a, b) => b.costSaved - a.costSaved);

    return entries;
  }

  private captureStatsSnapshot(): void {
    this.stats()
      .then((s) => {
        this.analytics.capture('stats_snapshot', {
          llm_hits: s.llm.hits,
          llm_misses: s.llm.misses,
          llm_hit_rate: s.llm.hitRate,
          tool_hits: s.tool.hits,
          tool_misses: s.tool.misses,
          tool_hit_rate: s.tool.hitRate,
          session_reads: s.session.reads,
          session_writes: s.session.writes,
          cost_saved_micros: s.costSavedMicros,
          tool_count: Object.keys(s.perTool).length,
        });
      })
      .catch(() => {});
  }

  async shutdown(): Promise<void> {
    this.shutdownCalled = true;
    if (this.statsTimer) {
      clearInterval(this.statsTimer);
      this.statsTimer = undefined;
    }
    await this.analytics.shutdown();
  }

  async flush(): Promise<void> {
    // Escape cache name in case it contains glob metacharacters
    const pattern = `${escapeGlobPattern(this.name)}:*`;

    try {
      await clusterScan(this.client, pattern, async (keys, nodeClient) => {
        // Use a pipeline of individual DEL commands — multi-key DEL causes
        // CROSSSLOT errors in cluster mode when keys span different hash slots.
        const pipeline = nodeClient.pipeline();
        for (const key of keys) pipeline.del(key);
        let delResults: Array<[Error | null, number]>;
        try {
          delResults = await pipeline.exec() as Array<[Error | null, number]>;
        } catch (err) {
          throw new ValkeyCommandError('DEL', err);
        }
        for (const [err] of delResults) {
          if (err) throw new ValkeyCommandError('DEL', err);
        }
      });
    } finally {
      // Always reset in-memory state even on partial failure — leaving stale
      // sessions or policies after some keys were deleted would be worse than
      // resetting eagerly and requiring a reload on the next operation.
      this.session.resetTracker();
      this.tool.resetPolicies();
      this.analytics.capture('cache_flush');
    }
  }
}

File: packages/agent-cache-py/betterdb_agent_cache/agent_cache.py
from __future__ import annotations

import asyncio
import re
from datetime import datetime, timezone
from importlib.metadata import PackageNotFoundError, version as _pkg_version
from typing import Any

from .analytics import NOOP_ANALYTICS, Analytics, create_analytics
from .cluster import cluster_scan
from .default_cost_table import DEFAULT_COST_TABLE
from .discovery import BuildAgentMetadataInput, DiscoveryManager, build_agent_metadata
from .errors import AgentCacheUsageError, ValkeyCommandError
from .telemetry import create_telemetry
from .tiers.llm_cache import LlmCache, LlmCacheConfig
from .tiers.session_store import SessionStore, SessionStoreConfig
from .tiers.tool_cache import ToolCache, ToolCacheConfig
from .types import (
    AgentCacheOptions,
    AgentCacheStats,
    ModelCost,
    SessionStats,
    TierStats,
    ToolEffectivenessEntry,
    ToolStats,
)
from .utils import escape_glob_pattern

try:
    _PACKAGE_VERSION = _pkg_version("betterdb-agent-cache")
except PackageNotFoundError:
    _PACKAGE_VERSION = "0.0.0"


class AgentCache:
    """Multi-tier exact-match cache for AI agent workloads backed by Valkey.

    Tiers:
    - ``llm``     — LLM response cache (check / store / store_multipart)
    - ``tool``    — Tool call result cache (check / store / set_policy)
    - ``session`` — Session state store (get / set / get_all / destroy_thread)
    """

    def __init__(self, options: AgentCacheOptions) -> None:
        self._client = options.client
        self._name = options.name
        self._stats_key = f"{self._name}:__stats"
        self._default_ttl = options.default_ttl
        self._tool_tier_defaults = options.tier_defaults.get("tool", None)
        self._cost_table = options.cost_table
        self._use_default_cost_table = options.use_default_cost_table
        self._analytics_opts = options.analytics
        self._analytics: Analytics = NOOP_ANALYTICS
        self._stats_task: asyncio.Task[None] | None = None
        self._config_refresh_task: asyncio.Task[None] | None = None
        self._background_tasks: set[asyncio.Task[None]] = set()
        self._shutdown = False
        self._started_at_iso = datetime.now(timezone.utc).isoformat()
        self._discovery: DiscoveryManager | None = None
        self._discovery_task: asyncio.Task[None] | None = None
        self._discovery_error: Exception | None = None

        use_default = options.use_default_cost_table
        effective_cost_table: dict[str, ModelCost] | None
        if use_default:
            effective_cost_table = {**DEFAULT_COST_TABLE, **(options.cost_table or {})}
        else:
            effective_cost_table = options.cost_table or None

        telemetry = create_telemetry(
            prefix=options.telemetry.metrics_prefix,
            tracer_name=options.telemetry.tracer_name,
            registry=options.telemetry.registry,
        )
        self._telemetry = telemetry

        refresh = options.config_refresh
        self._config_refresh_enabled = refresh.enabled
        self._config_refresh_interval_s = max(1.0, refresh.interval_ms / 1000)

        self.llm = LlmCache(LlmCacheConfig(
            client=self._client,
            name=self._name,
            default_ttl=options.default_ttl,
            tier_ttl=options.tier_defaults.get("llm", None) and options.tier_defaults["llm"].ttl,
            cost_table=effective_cost_table,
            telemetry=telemetry,
            stats_key=self._stats_key,
        ))

        self.tool = ToolCache(ToolCacheConfig(
            client=self._client,
            name=self._name,
            default_ttl=options.default_ttl,
            tier_ttl=options.tier_defaults.get("tool", None) and options.tier_defaults["tool"].ttl,
            telemetry=telemetry,
            stats_key=self._stats_key,
        ))

        self.session = SessionStore(SessionStoreConfig(
            client=self._client,
            name=self._name,
            default_ttl=options.default_ttl,
            tier_ttl=(
                options.tier_defaults.get("session", None)
                and options.tier_defaults["session"].ttl
            ),
            telemetry=telemetry,
            stats_key=self._stats_key,
        ))

        # Fire-and-forget: start config refresh loop, initialise analytics,
        # and register the discovery marker.
        # Uses get_running_loop() so this is a no-op when AgentCache is created
        # outside an async context.
        try:
            loop = asyncio.get_running_loop()
            if self._config_refresh_enabled:
                t = loop.create_task(self._config_refresh_loop())
                self._config_refresh_task = t
                self._background_tasks.add(t)
                t.add_done_callback(self._background_tasks.discard)
            t2 = loop.create_task(self._init_analytics_safe())
            self._background_tasks.add(t2)
            t2.add_done_callback(self._background_tasks.discard)
            t3 = loop.create_task(self._register_discovery(options))
            self._discovery_task = t3
            self._background_tasks.add(t3)
            t3.add_done_callback(self._background_tasks.discard)
        except RuntimeError:
            pass

    async def _config_refresh_loop(self) -> None:
        """Periodically refresh tool policies from Valkey.

        First refresh fires immediately (before the first sleep) so a process
        started right after a proposal is applied picks up the change without
        waiting a full interval. Subsequent refreshes fire every
        ``config_refresh_interval_s`` seconds.
        """
        try:
            while not self._shutdown:
                ok = await self.tool.refresh_policies()
                if not ok:
                    self._telemetry.metrics.config_refresh_failed.labels(self._name).inc()
                await asyncio.sleep(self._config_refresh_interval_s)
        except asyncio.CancelledError:
            pass

    async def _init_analytics_safe(self) -> None:
        try:
            opts = self._analytics_opts
            analytics = await create_analytics(disabled=opts.disabled)
            if self._shutdown:
                await analytics.shutdown()
                return
            self._analytics = analytics
            await analytics.init(self._client, self._name, {
                "default_ttl": self._default_ttl,
                "has_cost_table": bool(self._cost_table),
                "uses_default_cost_table": self._use_default_cost_table,
            })
            if not self._shutdown and opts.stats_interval_s > 0 and self._analytics is not NOOP_ANALYTICS:
                self._stats_task = asyncio.create_task(self._stats_loop(opts.stats_interval_s))
        except Exception:
            pass

    async def _register_discovery(self, options: AgentCacheOptions) -> None:
        disc_opts = options.discovery
        if not disc_opts.enabled:
            return

        include_tool_policies = disc_opts.include_tool_policies

        def build_metadata() -> dict:
            return build_agent_metadata(BuildAgentMetadataInput(
                name=self._name,
                version=_PACKAGE_VERSION,
                tiers={
                    'llm': {'ttl': options.tier_defaults.get('llm', None) and options.tier_defaults['llm'].ttl},
                    'tool': {'ttl': options.tier_defaults.get('tool', None) and options.tier_defaults['tool'].ttl},
                    'session': {'ttl': options.tier_defaults.get('session', None) and options.tier_defaults['session'].ttl},
                },
                default_ttl=self._default_ttl,
                tool_policy_names=self.tool.list_policy_names() if include_tool_policies else [],
                has_cost_table=bool(self._cost_table),
                uses_default_cost_table=self._use_default_cost_table,
                started_at=self._started_at_iso,
                include_tool_policies=include_tool_policies,
            ))

        manager = DiscoveryManager(
            client=self._client,
            name=self._name,
            build_metadata=build_metadata,
            heartbeat_interval_s=disc_opts.heartbeat_interval_ms / 1000.0,
            on_write_failed=lambda: self._telemetry.metrics.discovery_write_failed.labels(self._name).inc(),
        )

        try:
            await manager.register()
            if self._shutdown:
                await manager.stop(delete_heartbeat=True)
                return
            self._discovery = manager
        except AgentCacheUsageError as exc:
            self._discovery_error = exc
        except Exception as exc:
            self._discovery_error = exc

    async def ensure_discovery_ready(self) -> None:
        """Wait for the background discovery registration to complete.

        Raises the stored error if registration failed (e.g. cross-type
        collision).
        """
        if self._discovery_error is not None:
            raise self._discovery_error
        # Await only the one-shot discovery task, not _background_tasks which
        # includes the infinite config-refresh loop that never completes on its own.
        if self._discovery_task is not None and not self._discovery_task.done():
            await asyncio.gather(self._discovery_task, return_exceptions=True)
        if self._discovery_error is not None:
            raise self._discovery_error

    async def _stats_loop(self, interval_s: float) -> None:
        while not self._shutdown:
            try:
                await asyncio.sleep(interval_s)
                if self._shutdown:
                    break
                s = await self.stats()
                self._analytics.capture("stats_snapshot", {
                    "llm_hits": s.llm.hits,
                    "llm_misses": s.llm.misses,
                    "llm_hit_rate": s.llm.hit_rate,
                    "tool_hits": s.tool.hits,
                    "tool_misses": s.tool.misses,
                    "tool_hit_rate": s.tool.hit_rate,
                    "session_reads": s.session.reads,
                    "session_writes": s.session.writes,
                    "cost_saved_micros": s.cost_saved_micros,
                    "tool_count": len(s.per_tool),
                })
            except asyncio.CancelledError:
                break
            except Exception:
                pass

    async def stats(self) -> AgentCacheStats:
        try:
            raw: dict[bytes | str, bytes | str] = await self._client.hgetall(self._stats_key) or {}
        except Exception as exc:
            raise ValkeyCommandError("HGETALL", exc) from exc

        def decode(v: bytes | str) -> str:
            return v.decode() if isinstance(v, bytes) else v

        def get_int(field: str) -> int:
            val = raw.get(field) or raw.get(field.encode())
            return int(decode(val)) if val else 0

        llm_hits = get_int("llm:hits")
        llm_misses = get_int("llm:misses")
        tool_hits = get_int("tool:hits")
        tool_misses = get_int("tool:misses")

        per_tool: dict[str, ToolStats] = {}
        pattern = re.compile(r"^tool:([^:]+):(hits|misses|cost_saved_micros)$")

        for raw_key, raw_val in raw.items():
            key = decode(raw_key)
            match = pattern.match(key)
            if not match:
                continue
            tool_name, stat_type = match.group(1), match.group(2)
            num = int(decode(raw_val))
            if tool_name not in per_tool:
                policy = self.tool.get_policy(tool_name)
                per_tool[tool_name] = ToolStats(
                    hits=0, misses=0, ttl=policy.ttl if policy else None, cost_saved_micros=0
                )
            entry = per_tool[tool_name]
            if stat_type == "hits":
                entry.hits = num
            elif stat_type == "misses":
                entry.misses = num
            elif stat_type == "cost_saved_micros":
                entry.cost_saved_micros = num

        return AgentCacheStats(
            llm=TierStats(hits=llm_hits, misses=llm_misses),
            tool=TierStats(hits=tool_hits, misses=tool_misses),
            session=SessionStats(reads=get_int("session:reads"), writes=get_int("session:writes")),
            cost_saved_micros=get_int("cost_saved_micros"),
            per_tool=per_tool,
        )

    async def tool_effectiveness(self) -> list[ToolEffectivenessEntry]:
        stats = await self.stats()
        entries: list[ToolEffectivenessEntry] = []

        tool_tier = self._tool_tier_defaults
        tool_tier_ttl = tool_tier.ttl if tool_tier else None

        for tool_name, tool_stats in stats.per_tool.items():
            cost_saved = tool_stats.cost_saved_micros / 1_000_000
            policy = self.tool.get_policy(tool_name)
            effective_ttl = (policy.ttl if policy else None) or tool_tier_ttl or self._default_ttl

            if tool_stats.hit_rate > 0.8:
                recommendation = (
                    "increase_ttl"
                    if effective_ttl is not None and effective_ttl < 3600
                    else "optimal"
                )
            elif tool_stats.hit_rate >= 0.4:
                recommendation = "optimal"
            else:
                recommendation = "decrease_ttl_or_disable"

            entries.append(ToolEffectivenessEntry(
                tool=tool_name,
                hit_rate=tool_stats.hit_rate,
                cost_saved=cost_saved,
                recommendation=recommendation,
            ))

        entries.sort(key=lambda e: e.cost_saved, reverse=True)
        return entries

    async def flush(self) -> None:
        pattern = f"{escape_glob_pattern(self._name)}:*"
        try:
            async def on_keys(keys: list[str], client: Any) -> None:
                pipe = client.pipeline(transaction=False)
                for k in keys:
                    pipe.delete(k)
                try:
                    results = await pipe.execute()
                except Exception as exc:
                    raise ValkeyCommandError("DEL", exc) from exc
                for err in results:
                    if isinstance(err, Exception):
                        raise ValkeyCommandError("DEL", err)

            await cluster_scan(self._client, pattern, on_keys)
        finally:
            self.session.reset_tracker()
            self.tool.reset_policies()
            self._analytics.capture("cache_flush")

    async def shutdown(self) -> None:
        self._shutdown = True
        if self._config_refresh_task is not None:
            self._config_refresh_task.cancel()
            self._config_refresh_task = None
        if self._stats_task is not None:
            self._stats_task.cancel()
            self._stats_task = None
        if self._discovery is not None:
            await self._discovery.stop(delete_heartbeat=True)
            self._discovery = None
        await self._analytics.shutdown()
